Allow local development

yawaramin opened the pull request:
Allow specifying an origin like `http://localhost:8080` for dev and testing purposes.

Allow the demo to validate macOS Touch ID non-increasing counters which are always set to 0 (zero).

yawaramin committed:
Tell git to ignore build directory

yawaramin committed:
Allow http protocol for local development

When the 'origin' starts with `http://localhost:`, allow the WebAuthn
value to be created so that we can test it during local development.

yawaramin committed:
Allow signature counter zero (0)

Due to certain authenticators like Touch ID on macOS always returning
zero (0) for the signature counter.

reynir commented on src/webauthn.mli:
```diff
@@ -30,8 +30,11 @@
+    meet the specification: schema must be https, the host must be a valid
+    hostname. An optional port is supported: https://example.com:4444
+
+    For local development purposes, it is allowed to specify an origin like
+    [http://localhost:PORT] where [PORT] is a port number. *)
```
Maybe:
```suggestion
    For local development purposes, it is allowed to specify an origin with
    scheme [http] {e and} host [localhost]. *)
```
This follows more closely the spec (https://w3c.github.io/webauthn/#rp-id) and doesn't imply the port is required.

reynir commented on bin/webauthn_demo.ml:
```diff
@@ -35,7 +35,16 @@ let counters = KhPubHashtbl.create 7
+    | Some counter' ->
+      Int32.unsigned_compare counter counter' > 0
+
+      (* As per https://simplewebauthn.dev/docs/packages/server#3-post-registration-responsibilities
+
+        'It's also not unexpected for certain high profile authenticators, like
+        Touch ID on macOS, to always return 0 (zero) for the signature counter.
+        In this case there is nothing an RP can really do to detect a cloned
+        authenticator....' *)
```
Spec wise it doesn't mandate what exactly we do if the counter does not appear to be increasing. Perhaps we should revisit this and always authenticate, but modify the flash message to also say something if the counter is not increasing or if the counter is zero.

reynir commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+_build
+
```
Personally I put this in the global gitignore https://git-scm.com/docs/gitignore (e.g. on my machine `~/.config/git/ignore`).

reynir reviewed (commented):
Thanks a lot! I will make some minor changes and then it looks good to merge to me.

reynir committed:
Use OCaml>=4.13.0, allow http://localhost origin

We should allow the http://localhost origin, i.e. with the implicit
standard port.

reynir committed:
Reword doc about localhost origin

This follows more closely the specification and does not imply that an
explicit port is required.

reynir committed:
Reference the spec instead of a 3rd party impl.
